### PR TITLE
EKF2/EKF3: Check if EKF origin exists before setting origin in non-GPS modes

### PR DESCRIPTION
**Pull request by hendjoshsr71**
This PR checks, at the EKF level, if the EKF origin exists before allowing the EKF origin to be set. This can occur in a mode without GPS such as visual odometry.

Issue found in #17766 as the origin was getting set via scripting.

**Master:**
The underlined points are where I reset the new EKF origin to be. The white circles are the beginning and end point after the reset. 

image

**Below shows the set_origin command being refused:**
Capture

### CODE REVIEW

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 1]
```diff
@@ -1358,9 +1358,11 @@ bool NavEKF3::setOriginLLH(const Location &loc)
     if (!core) {
         return false;
     }
-    if (sources.getPosXYSource() == AP_NavEKF_Source::SourceXY::GPS) {
-        // we don't allow setting of the EKF origin if using GPS to prevent
-        // accidental setting of EKF origin with invalid position or height
+    if ((sources.getPosXYSource() == AP_NavEKF_Source::SourceXY::GPS) || common_origin_valid) {
```
Comment (rmackay9) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp:1361]: my guess is that we should use "validOrigin" (the local boolean) instead of "common_origin_valid".

**Comment by rmackay9** [comment]
Hi, thanks for this.  It looks correct to me although as mentioned in a comment above, I think we should use the core specific, "validOrigin" variable instead of the front-end variable.

While looking at this I've uncovered other things that look odd:

- It looks like we have a bug in NavEKF3_core::setOrigin() in the case where inFlight is true because the core's EKF_origin.alt will not be the common EKF origin alt.
- It would be nice if we could combine setOriginLLH() and setOrigin() because they are so similar.  More code sharing or comments on how the two should be used would be nice

We shouldn't try and address these additional questions in this PR but I wanted to write them down for the record

**Comment by rmackay9** [comment]
I also think both setOriginLLH() and setOrigin() should return immediately (e..g reject setting the origin) if validOrigin is true.  perhaps setOrigin() should be changed to return a boolean or perhaps it should raise an internal error (a control_flow most likely).

I'm going to discuss this issue with Tridge and Paul Riseborough to decide how we should fix this issue I've found.

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp (hunk 1 above)
Comment (hendjoshsr71) [reply libraries/AP_NavEKF3/AP_NavEKF3.cpp:1361]: The reason I was using `common_origin_valid` is that at this level in `class NavEKF3`  `validOrigin` isn't available. It is part of the core level `class NavEKF3_core`. I guess we could loop through each core and check `validOrigin` but getting the common_origin_valid correct and using that at this higher level seems like the right thing to me still?

I agree the placement of where the cores are insured to have the same origin is odd. When looking at it It took me a bit to figure out how the cores were each getting set to the new the common_origin.

**Comment by hendjoshsr71** [comment]
> I also think both setOriginLLH() and setOrigin() should return immediately (e..g reject setting the origin) if validOrigin is true. perhaps setOrigin() should be changed to return a boolean or perhaps it should raise an internal error (a control_flow most likely).
> 
> I'm going to discuss this issue with Tridge and Paul Riseborough to decide how we should fix this issue I've found.

Yes, I was also thinking about that. Having checks at each level for the origin seems worth any extra cost. I will add the check into the core level as well and look at the consequence of switching setOrigin() to bool.

**Comment by hendjoshsr71** [comment]
Added additional checks into `NavEKF3_core::setOriginLLH`. Used a flow_control  in `setOrigin` since bool doesn`t make sense in terms of how it gets called.

**Comment by rmackay9** [comment]
@hendjoshsr71,

I had a chat with Tridge about the bug (issue https://github.com/ArduPilot/ardupilot/issues/17789) and at this point I think we should add the solution to this PR because we're going to need to test it in either case.  So here's what I think we should do:

- move the "EKF_origin.alt += (int32_t)(100.0f * stateStruct.position.z);}" part out of setOrigin() to be just before this is called from within readGPS().
- remove setOrigin() completely and only leave setOriginLLH()
- maybe remove the source==gps from the top of setOriginLLH()

**Comment by hendjoshsr71** [comment]
@rmackay9

I think I have the altitude difference between cores fixed see testing below. 

I kept both functions as one is used publicly and the other privately. This way we can keep the conditions at this lower level preventing the origin being set from scripting/GCS if GPS is enabled for the public version which is for non-GPS modes only.

Master:
image


Here is after the PR:  Tested using the method you put in #17789
image

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp [hunk 2]
```diff
@@ -584,15 +584,26 @@ void NavEKF2_core::readGpsData()
 
             // see if we can get origin from frontend
             if (!validOrigin && frontend->common_origin_valid) {
-                setOrigin(frontend->common_EKF_origin);
+
+                if(!setOrigin(frontend->common_EKF_origin)) {
```
Comment (rmackay9) [inline libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp:588]: nitpick: can you add a space between the "if" and the bracket?  sorry, i know I'm being annoying..

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp [hunk 3]
```diff
@@ -594,10 +594,14 @@ void NavEKF2_core::readGpsData()
...
 
             // Read the GPS location in WGS-84 lat,long,height coordinates
-            const struct Location &gpsloc = gps.location();
+            Location gpsloc = gps.location();
 
             // Set the EKF origin and magnetic field declination if not previously set  and GPS checks have passed
             if (gpsGoodToAlign && !validOrigin) {
+                // if flying, correct for height change from takeoff so that the origin is at field elevation
+                if (inFlight) {
+                    gpsloc.alt += (int32_t)(100.0f * stateStruct.position.z);
```
Comment (rmackay9) [inline libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp:603]: I think unfortunately you're going to need to create another Location object called something like "gpsloc_fieldelevation" and only adjust its altitude.  The issue is just that gpsloc is consumed a little lower down in the method as well.

**Comment by rmackay9** [comment]
Looking pretty nice actually.  Txs very much for this.. one real issue to fix and then hopefully we can get Tridge and Paul to give their OK and we can merge it.  txs for this.

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp (hunk 2 above)
Comment (hendjoshsr71) [reply libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp]: fixed. no problem

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp (hunk 3 above)
Comment (hendjoshsr71) [reply libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp]: Fixed. Yup sorry about that one I should have checked better.

**Review by rmackay9 — APPROVED** [review]
This looks good to me.  I've re-tested in SITL and the EKF origin altitude is now fixed (see screenshot below) and also the origin cannot be set twice on either EKF2 or EKF3.  Thanks!

ekf-origin-fix-before-vs-after

File: libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp [hunk 4]
```diff
@@ -585,30 +585,28 @@ bool NavEKF3_core::assume_zero_sideslip(void) const
...
-void NavEKF3_core::setOrigin(const Location &loc)
+// sets the local NED origin using a LLH location (latitude, longitude, height)
+// returns false is the origin has already been set
+bool NavEKF3_core::setOrigin(const Location &loc)
 {
-    EKF_origin = loc;
-    // if flying, correct for height change from takeoff so that the origin is at field elevation
-    if (inFlight) {
-        EKF_origin.alt += (int32_t)(100.0f * stateStruct.position.z);
+    // if the origin is valid reject setting a new origin and set an error
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp:604]: Not it doesn't :-)

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp [hunk 5]
```diff
@@ -646,15 +646,30 @@ void NavEKF3_core::readGpsData()
 
             // see if we can get an origin from the frontend
             if (!validOrigin && frontend->common_origin_valid) {
-                setOrigin(frontend->common_EKF_origin);
+
+                if (!setOrigin(frontend->common_EKF_origin)) {
+                    // rejects setting the origin more than once and sets an error
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:651]:
```suggestion
                    // rejects setting the origin more than once and sets an error
```

Not it doesn't - similarly elsewhere

**Review by rmackay9 — APPROVED** [review]

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp (hunk 4 above)
Comment (hendjoshsr71) [reply libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp]: fixed :)

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp (hunk 5 above)
Comment (hendjoshsr71) [reply libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp]: fixed :)

**Pull request merged**